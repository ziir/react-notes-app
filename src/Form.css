.Form {
}

.Form-fieldset {
  display: flex;
  flex-direction: column;
  border: none;
}

.Form-label {
  margin-bottom: var(--spacing-xsmall);
}

.Form-input {
  padding: var(--spacing-xsmall);
  min-height: var(--spacing-large);
  border: 1px solid var(--color-gray-dark);
}

.Form-textarea {
  resize: vertical;
}
