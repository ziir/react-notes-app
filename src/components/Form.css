.Form {
  flex-grow: 1;
}

.Form-fieldset {
  display: flex;
  flex-direction: column;
  border: none;
  padding: 0;
  margin: 0;
  margin-bottom: var(--spacing);
}

.Form-label {
  margin-bottom: var(--spacing-xsmall);
}

.Form-input {
  padding: var(--spacing-xsmall);
  min-height: var(--spacing-large);
  width: 100%;
  border: 1px solid var(--color-gray-dark);
}

.Form-textarea {
  resize: vertical;
}

.Form-actions {
  display: flex;
  justify-content: space-between;
}

.Form-actions > * {
  font-size: 1rem;
  line-height: 1rem;
}

.Form-cancel-link {
  background-color: white;
}

.Form-cancel-link:hover,
.Form-cancel-link:focus {
  box-shadow: inset 0 0 0 var(--spacing-xsmall) rgba(0, 0, 0, 0.1);
  outline: none;
  filter: none;
}
