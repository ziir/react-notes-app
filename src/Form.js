import React from 'react';

import './Form.css';

async function handleSubmit(setSubmitting, setError, onSubmit, evt) {
  evt.preventDefault();
  setSubmitting(true);

  const formData = new FormData(evt.target);
  const item = {};
  for (let [key, value] of formData.entries()) {
    item[key] = value;
  }

  try {
    await onSubmit(item);
  } catch (err) {
    setError(err);
  } finally {
    setSubmitting(false);
  }
}

function Form({
  formId: id,
  title = '',
  content = '',
  onSubmit,
  children: renderAdjacent = null,
}) {
  const [submitting, setSubmitting] = React.useState(false);
  const [error, setError] = React.useState(null);

  return (
    <>
      {error && <p className="Form-Error">{error.toString()}</p>}
      <form
        id={id}
        className="Form"
        onSubmit={handleSubmit.bind(null, setSubmitting, setError, onSubmit)}
      >
        <fieldset className="Form-fieldset">
          <label className="Form-label" htmlFor="title-input">
            Title
          </label>
          <input
            required
            autoFocus
            type="text"
            name="title"
            id="title-input"
            disabled={submitting}
            className="Form-input"
            placeholder="Enter title here..."
            defaultValue={title}
          />
        </fieldset>
        <fieldset className="Form-fieldset">
          <label className="Form-label" htmlFor="content-input">
            Content
          </label>
          <textarea
            id="content-input"
            name="content"
            rows="10"
            placeholder="Enter markdown content here..."
            className="Form-input Form-textarea"
            disabled={submitting}
            defaultValue={content}
          ></textarea>
        </fieldset>
      </form>
      {typeof renderAdjacent === 'function' && renderAdjacent({ submitting })}
    </>
  );
}

export default Form;
