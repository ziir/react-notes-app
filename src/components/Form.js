import React from 'react';

import { Link } from '../modules/Router/Link';
import './Form.css';

async function handleSubmit(setSubmitting, setError, onSubmit, evt) {
  evt.preventDefault();
  setSubmitting(true);

  const formData = new FormData(evt.target);
  const item = {};
  for (let [key, value] of formData.entries()) {
    item[key] = value;
  }

  try {
    await onSubmit(item);
  } catch (err) {
    setError(err);
  } finally {
    setSubmitting(false);
  }
}

function Form({
  id = null,
  formId,
  title = '',
  content = '',
  onSubmit,
  children: renderAdjacent = null,
}) {
  const [submitting, setSubmitting] = React.useState(false);
  const [error, setError] = React.useState(null);

  return (
    <>
      {error && <p className="Form-Error">{error.toString()}</p>}
      <form
        id={formId}
        className="Form"
        onSubmit={handleSubmit.bind(null, setSubmitting, setError, onSubmit)}
      >
        <fieldset className="Form-fieldset">
          <label className="Form-label" htmlFor="title-input">
            Title
          </label>
          <input
            required
            autoFocus
            type="text"
            name="title"
            id="title-input"
            disabled={submitting}
            className="Form-input"
            placeholder="Enter title here..."
            defaultValue={title}
          />
        </fieldset>
        <fieldset className="Form-fieldset">
          <label className="Form-label" htmlFor="content-input">
            Content
          </label>
          <textarea
            id="content-input"
            name="content"
            rows="10"
            placeholder="Enter markdown content here..."
            className="Form-input Form-textarea"
            disabled={submitting}
            defaultValue={content}
          ></textarea>
        </fieldset>
      </form>
      <div className="Form-actions">
        <Link
          className="Form-cancel-link"
          disabled={submitting}
          to={id ? `/view/${id}` : '/'}
        >
          Cancel
        </Link>
        {typeof renderAdjacent === 'function' && renderAdjacent({ submitting })}
      </div>
    </>
  );
}

export default Form;
