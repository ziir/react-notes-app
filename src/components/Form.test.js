import React from 'react';
import { render, fireEvent, act } from '@testing-library/react';
import Form from './Form';

const item = {
  id: '1',
  title: 'Blog Post Idea',
  content: 'Lorem Ipsum Dolor Sit Amet',
};
const baseProps = {
  formId: 'test-form',
  onSubmit: () => {},
};

describe('Creation mode (minimal props)', () => {
  const defaultProps = { ...baseProps };

  test('renders a form element with the provided formId', () => {
    const { getByRole } = render(<Form {...defaultProps} />);
    const form = getByRole('form');
    expect(form.id).toBe('test-form');
  });

  test('renders a title text input', () => {
    const { getByLabelText } = render(<Form {...defaultProps} />);
    const input = getByLabelText(/title/i);
    expect(input.type).toBe('text');
  });

  test('renders a content textarea', () => {
    const { getByLabelText } = render(<Form {...defaultProps} />);
    const input = getByLabelText(/content/i);
    expect(input.nodeName).toBe('TEXTAREA');
  });

  test('renders a cancel link', () => {
    const { getByText } = render(<Form {...defaultProps} />);
    const input = getByText(/cancel/i);
    expect(input.getAttribute('href')).toBe('/');
  });

  test('calls a renderProp with the `submitting` state (initial)', () => {
    const props = { ...defaultProps, children: jest.fn() };
    render(<Form {...props} />);
    expect(props.children).toHaveBeenCalledWith({ submitting: false });
  });

  test('calls a renderProp with the `submitting` state (with submit)', async () => {
    const promise = Promise.resolve();
    const props = {
      ...defaultProps,
      onSubmit: () => {},
      children: jest.fn(),
    };
    const { getByRole } = render(<Form {...props} />);
    const form = getByRole('form');
    fireEvent.submit(form);
    await act(() => promise);
    expect(props.children).toMatchSnapshot();
  });

  test('calls onSubmit prop when submitted', async () => {
    const promise = Promise.resolve();
    const props = {
      ...defaultProps,
      onSubmit: jest.fn(() => promise),
    };

    const { getByRole, getByLabelText } = render(<Form {...props} />);

    const title = getByLabelText(/title/i);
    fireEvent.change(title, { target: { value: item.title } });

    const content = getByLabelText(/content/i);
    fireEvent.change(content, { target: { value: item.content } });

    const form = getByRole('form');
    fireEvent.submit(form);

    expect(props.onSubmit).toHaveBeenCalledWith({
      title: item.title,
      content: item.content,
    });
    await act(() => promise);
  });
});

describe('Edition mode (full props)', () => {
  const defaultProps = { ...baseProps, ...item };

  test('renders & pre-fill a title text input', () => {
    const { getByLabelText } = render(<Form {...defaultProps} />);
    const input = getByLabelText(/title/i);
    expect(input.value).toBe(item.title);
  });

  test('renders & pre-fill a content textarea', () => {
    const { getByLabelText } = render(<Form {...defaultProps} />);
    const input = getByLabelText(/content/i);
    expect(input.value).toBe(item.content);
  });

  test('renders a cancel link to the view mode', () => {
    const { getByText } = render(<Form {...defaultProps} />);
    const input = getByText(/cancel/i);
    expect(input.getAttribute('href')).toBe(`/view/${item.id}`);
  });
});
