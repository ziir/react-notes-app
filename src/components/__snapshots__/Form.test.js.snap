// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Creation mode (minimal props) calls a renderProp with the \`submitting\` state (with submit) 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "submitting": false,
      },
    ],
    Array [
      Object {
        "submitting": true,
      },
    ],
    Array [
      Object {
        "submitting": false,
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
